File: internal/evidence/evidence.go
// Package evidence ties a decision to the upstream state it was computed
// from: what each check asked the upstream and what the upstream answered,
// carried on the context so httpx and the caches can record and replay it.
package evidence

import (
	"context"
	"encoding/json"
	"slices"
	"sync"
)

// Call is one completed upstream request.
type Call struct {
	Method string `json:"method"`
	// Path is the request path as sent. Never the query string, which may
	// carry user data or a token.
	Path string `json:"path"`
	// Host is set only when the call went to a host other than the
	// connection's own base URL: some vendors spread an API over several
	// hosts, and a path alone would not say which answered.
	Host   string `json:"host,omitempty"`
	Status int    `json:"status"`
	// ETag is the response's ETag header, when it sent one.
	ETag string `json:"etag,omitempty"`
	// SHA256 is the hex SHA-256 of the response body, when there was no
	// ETag and the body was not empty.
	SHA256 string `json:"sha256,omitempty"`
	// Cached marks a call that was not made for this check: its result was
	// served from a stored cache entry (the identity cache, or one an
	// integration keeps), and this is the evidence recorded when the call
	// was made.
	Cached bool `json:"cached,omitempty"`
	// Shared marks a call made by a concurrent check whose lookup this
	// check joined: the response was live during this check, but this
	// check did not ask for it itself.
	Shared bool `json:"shared,omitempty"`
}

// MaxCalls bounds the calls one Evidence reports, so a paginated lookup
// cannot grow a log line without limit. Calls made for the check take
// precedence over replayed ones, and among those the latest are kept: the
// call that decided a check is the last one it made.
const MaxCalls = 100

// Origin says how a check came by a call's response.
type Origin int

const (
	// Own: the check made the call.
	Own Origin = iota
	// Shared: a concurrent check made the call and this check joined it.
	Shared
	// Cached: the call was made earlier and its result served from a cache.
	Cached
)

// Evidence is the calls a decision was based on and, for each, what
// identifies the version of the response (an ETag when the upstream sent
// one, otherwise a hash of the body). It goes into the decision log so an
// auditor can tell what the upstream system said when hallpass answered,
// not only the outcome. Most vendor APIs expose no policy version, so this
// is the closest thing available.
//
// An Evidence is immutable once built. The calls a check got from a cache
// are held by reference to the evidence recorded when they were made, so a
// cached lookup's pages are stored once however many decisions replay
// them; Calls flattens them on demand, marked by origin.
type Evidence struct {
	items []item
	// truncated is set when own calls were dropped at the cap.
	truncated bool
}

// item is one own call or a reference to another Evidence's calls.
type item struct {
	call Call
	src  *Evidence
	by   Origin
}

// Of builds an Evidence of the given calls, as recorded.
func Of(calls ...Call) *Evidence {
	if len(calls) == 0 {
		return nil
	}
	e := &Evidence{items: make([]item, 0, len(calls))}
	for _, c := range calls {
		e.items = append(e.items, item{call: c})
	}
	return e
}

// Calls returns the calls in the order they completed, marked by how the
// check came by them, at most MaxCalls of them. Past the cap the oldest
// calls are dropped, cached ones first. A nil Evidence has none.
func (e *Evidence) Calls() []Call {
	calls, _ := e.flatten()
	return calls
}

// Truncated reports whether calls were dropped at the cap.
func (e *Evidence) Truncated() bool {
	_, truncated := e.flatten()
	return truncated
}

func (e *Evidence) flatten() ([]Call, bool) {
	if e == nil {
		return nil, false
	}
	var out []Call
	truncated := e.walk(&out, Own)
	for len(out) > MaxCalls {
		truncated = true
		i := slices.IndexFunc(out, func(c Call) bool { return c.Cached })
		if i < 0 {
			i = 0
		}
		out = slices.Delete(out, i, i+1)
	}
	return out, truncated
}

// walk appends e's calls to out with by applied, and reports whether any
// segment on the way was truncated.
func (e *Evidence) walk(out *[]Call, by Origin) bool {
	if e == nil {
		return false
	}
	truncated := e.truncated
	for _, it := range e.items {
		if it.src != nil {
			inner := it.by
			if by == Cached {
				inner = Cached
			} else if by == Shared && inner == Own {
				inner = Shared
			}
			if it.src.walk(out, inner) {
				truncated = true
			}
			continue
		}
		c := it.call
		switch by {
		case Cached:
			c.Cached, c.Shared = true, false
		case Shared:
			c.Shared = !c.Cached
		}
		*out = append(*out, c)
	}
	return truncated
}

// AsCached returns e's calls as served from a cache, for a decision served
// from the decision cache. A nil e gives nil.
func (e *Evidence) AsCached() *Evidence {
	if e == nil {
		return nil
	}
	return &Evidence{items: []item{{src: e, by: Cached}}}
}

// wire is the JSON shape: the flattened calls.
type wire struct {
	Upstream  []Call `json:"upstream"`
	Truncated bool   `json:"truncated,omitempty"`
}

// MarshalJSON writes {"upstream": [...], "truncated": true}.
func (e *Evidence) MarshalJSON() ([]byte, error) {
	calls, truncated := e.flatten()
	if calls == nil {
		calls = []Call{}
	}
	return json.Marshal(wire{Upstream: calls, Truncated: truncated})
}

// UnmarshalJSON reads what MarshalJSON wrote, keeping each call's marks.
func (e *Evidence) UnmarshalJSON(b []byte) error {
	var w wire
	if err := json.Unmarshal(b, &w); err != nil {
		return err
	}
	*e = Evidence{truncated: w.Truncated}
	for _, c := range w.Upstream {
		e.items = append(e.items, item{call: c})
	}
	return nil
}

// Recorder collects the evidence of one check. It is safe for concurrent
// use: a shared fill may still be running on a detached context after the
// check that started it has returned.
//
// The engine gives every check a Recorder through its context; httpx
// records each response it returns on it, and cache.TTL.Do adds the
// evidence of a fill, by reference, marked by how the check came by it.
type Recorder struct {
	mu  sync.Mutex
	ev  Evidence
	own int
}

// Record adds one call the check made. A nil Recorder records nothing.
// Past MaxCalls own calls the oldest own call is dropped: the calls the
// check made last are the ones that decided it.
func (r *Recorder) Record(c Call) {
	if r == nil {
		return
	}
	r.mu.Lock()
	defer r.mu.Unlock()
	if r.own >= MaxCalls {
		r.ev.truncated = true
		i := slices.IndexFunc(r.ev.items, func(it item) bool { return it.src == nil })
		r.ev.items = slices.Delete(r.ev.items, i, i+1)
		r.own--
	}
	r.ev.items = append(r.ev.items, item{call: c})
	r.own++
}

// Add adds ev's calls, by reference, marked by how this check came by
// them. A nil ev adds nothing.
func (r *Recorder) Add(ev *Evidence, by Origin) {
	if r == nil || ev == nil {
		return
	}
	r.mu.Lock()
	defer r.mu.Unlock()
	r.ev.items = append(r.ev.items, item{src: ev, by: by})
}

// Evidence returns a snapshot of what was recorded, or nil when nothing
// was. The snapshot is not affected by later calls to Record or Add.
func (r *Recorder) Evidence() *Evidence {
	if r == nil {
		return nil
	}
	r.mu.Lock()
	defer r.mu.Unlock()
	if len(r.ev.items) == 0 {
		return nil
	}
	return &Evidence{items: slices.Clone(r.ev.items), truncated: r.ev.truncated}
}

type recorderKey struct{}

// WithRecorder returns a context carrying a new Recorder. The engine sets
// one up for every check.
func WithRecorder(ctx context.Context) (context.Context, *Recorder) {
	r := &Recorder{}
	return context.WithValue(ctx, recorderKey{}, r), r
}

// WithoutRecorder returns a context whose RecorderFrom is nil. httpx runs
// Auth funcs on it and authx fetches tokens on it: a token exchange is not
// what a decision was based on, and its response carries the credential.
func WithoutRecorder(ctx context.Context) context.Context {
	if ctx.Value(recorderKey{}) == nil {
		return ctx
	}
	return context.WithValue(ctx, recorderKey{}, (*Recorder)(nil))
}

// RecorderFrom returns the context's Recorder, or nil when the context has
// none or recording is suppressed. Every Recorder method accepts nil.
func RecorderFrom(ctx context.Context) *Recorder {
	r, _ := ctx.Value(recorderKey{}).(*Recorder)
	return r
}

type freshKey struct{}

// WithFresh marks a context as belonging to a fresh check: every cache.TTL
// consulted under it looks the value up again instead of serving a cached
// or in-flight one, and replaces its entry with the answer. The engine sets
// it for a request with "fresh": true.
func WithFresh(ctx context.Context) context.Context {
	return context.WithValue(ctx, freshKey{}, true)
}

// Fresh reports whether ctx belongs to a fresh check.
func Fresh(ctx context.Context) bool {
	v, _ := ctx.Value(freshKey{}).(bool)
	return v
}

File: hallpass-py/src/hallpass/core/evidence.py
"""What each check asked the upstream and what the upstream answered.

The engine gives every check a Recorder through its Context; the HTTP
client records each response it returns on it, and a cache fill adds the
evidence of its lookup, by reference, marked by how the check came by it
(its own call, a concurrent check's it joined, or a cached one). It also
holds the fresh mark, so the packages on a check's way share them without
depending on each other.
"""

from __future__ import annotations

import threading
from dataclasses import dataclass, replace
from enum import IntEnum
from typing import Any

from hallpass.core.context import Context, with_value

__all__ = [
    "MAX_CALLS",
    "Call",
    "Evidence",
    "Origin",
    "Recorder",
    "fresh",
    "of",
    "recorder_from",
    "with_fresh",
    "with_recorder",
    "without_recorder",
]

# The calls one Evidence reports, at most, so a paginated lookup cannot
# grow a log line without limit. Calls made for the check take precedence
# over replayed ones, and among those the latest are kept.
MAX_CALLS = 100


@dataclass(frozen=True)
class Call:
    """One completed upstream request."""

    method: str = ""
    # The request path as sent. Never the query string.
    path: str = ""
    # Set when the call went to a host other than the client's base URL.
    host: str = ""
    status: int = 0
    # The response's ETag header, when it sent one.
    etag: str = ""
    # The hex SHA-256 of the response body, when there was no ETag and the
    # body was not empty.
    sha256: str = ""
    # Not made for this check: served from a stored cache entry.
    cached: bool = False
    # Made by a concurrent check whose lookup this check joined.
    shared: bool = False

    def to_json(self) -> dict[str, Any]:
        out: dict[str, Any] = {"method": self.method, "path": self.path}
        if self.host:
            out["host"] = self.host
        out["status"] = self.status
        if self.etag:
            out["etag"] = self.etag
        if self.sha256:
            out["sha256"] = self.sha256
        if self.cached:
            out["cached"] = True
        if self.shared:
            out["shared"] = True
        return out

    @staticmethod
    def from_json(d: dict[str, Any]) -> Call:
        return Call(
            method=str(d.get("method", "")),
            path=str(d.get("path", "")),
            host=str(d.get("host", "")),
            status=int(d.get("status", 0)),
            etag=str(d.get("etag", "")),
            sha256=str(d.get("sha256", "")),
            cached=bool(d.get("cached", False)),
            shared=bool(d.get("shared", False)),
        )


class Origin(IntEnum):
    OWN = 0  # the check made the call
    SHARED = 1  # a concurrent check made it and this check joined it
    CACHED = 2  # made earlier and served from a cache


class _Item:
    __slots__ = ("by", "call", "src")

    def __init__(self, call: Call | None = None, src: Evidence | None = None, by: Origin = Origin.OWN) -> None:
        self.call = call
        self.src = src
        self.by = by


class Evidence:
    """The calls a decision was based on. Immutable once built.

    Calls a check got from a cache are held by reference to the evidence
    recorded when they were made, so a cached lookup's pages are stored
    once however many decisions replay them; calls() flattens on demand.
    """

    __slots__ = ("_cached", "_flat", "_items", "_lock", "_truncated")

    def __init__(self, items: list[_Item] | None = None, truncated: bool = False) -> None:
        self._items: list[_Item] = items or []
        self._truncated = truncated
        self._lock = threading.Lock()
        self._flat: tuple[list[Call], bool] | None = None
        self._cached: Evidence | None = None

    def calls(self) -> list[Call]:
        """The calls in completion order, marked by origin, at most
        MAX_CALLS; past the cap the oldest go, cached ones first."""
        return self._flatten()[0]

    def truncated(self) -> bool:
        return self._flatten()[1]

    def _flatten(self) -> tuple[list[Call], bool]:
        with self._lock:
            if self._flat is None:
                self._flat = self._flatten_once()
            return self._flat

    def _flatten_once(self) -> tuple[list[Call], bool]:
        total = 0
        cached = 0

        def count(c: Call) -> None:
            nonlocal total, cached
            total += 1
            if c.cached:
                cached += 1

        truncated = self._walk(Origin.OWN, count)
        drop_cached = drop_live = 0
        excess = total - MAX_CALLS
        if excess > 0:
            truncated = True
            drop_cached = min(excess, cached)
            drop_live = excess - drop_cached
        out: list[Call] = []

        def keep(c: Call) -> None:
            nonlocal drop_cached, drop_live
            if c.cached and drop_cached > 0:
                drop_cached -= 1
            elif not c.cached and drop_live > 0:
                drop_live -= 1
            else:
                out.append(c)

        self._walk(Origin.OWN, keep)
        return out, truncated

    def _walk(self, by: Origin, visit: Any) -> bool:
        truncated = self._truncated
        for it in self._items:
            if it.src is not None:
                inner = it.by
                if by == Origin.CACHED:
                    inner = Origin.CACHED
                elif by == Origin.SHARED and inner == Origin.OWN:
                    inner = Origin.SHARED
                if it.src._walk(inner, visit):
                    truncated = True
                continue
            c = it.call
            assert c is not None
            if by == Origin.CACHED:
                c = replace(c, cached=True, shared=False)
            elif by == Origin.SHARED:
                c = replace(c, shared=not c.cached)
            visit(c)
        return truncated

    def as_cached(self) -> Evidence:
        """This evidence's calls as served from a cache (the decision cache)."""
        with self._lock:
            if self._cached is None:
                self._cached = Evidence([_Item(src=self, by=Origin.CACHED)])
            return self._cached

    def to_json(self) -> dict[str, Any]:
        calls, truncated = self._flatten()
        out: dict[str, Any] = {"upstream": [c.to_json() for c in calls]}
        if truncated:
            out["truncated"] = True
        return out

    @staticmethod
    def from_json(d: dict[str, Any]) -> Evidence:
        return Evidence([_Item(call=Call.from_json(c)) for c in d.get("upstream") or []], bool(d.get("truncated", False)))


def of(*calls: Call) -> Evidence | None:
    """An Evidence of the given calls, as recorded."""
    if not calls:
        return None
    return Evidence([_Item(call=c) for c in calls])


def as_cached(ev: Evidence | None) -> Evidence | None:
    return None if ev is None else ev.as_cached()


class Recorder:
    """Collects the evidence of one check. Safe for concurrent use: a shared
    fill may still run after the check that started it has returned."""

    def __init__(self) -> None:
        self._lock = threading.Lock()
        self._items: list[_Item] = []
        self._truncated = False
        self._own = 0
        # When the oldest read added with add_at began, and the oldest a
        # fresh check would not tolerate; None when none.
        self._oldest: float | None = None
        self._strict: float | None = None

    def record(self, c: Call) -> None:
        """Add one call the check made. Past MAX_CALLS own calls the oldest
        own call is dropped: the calls made last decided the check."""
        with self._lock:
            if self._own >= MAX_CALLS:
                self._truncated = True
                for i, it in enumerate(self._items):
                    if it.src is None:
                        del self._items[i]
                        break
                self._own -= 1
            self._items.append(_Item(call=c))
            self._own += 1

    def add(self, ev: Evidence | None, by: Origin) -> None:
        """Add ev's calls by reference, marked by how the check came by them."""
        self.add_at(ev, by, None, None)

    def add_at(self, ev: Evidence | None, by: Origin, read_at: float | None, strict: float | None) -> None:
        """add for a read that began at read_at (a cache entry or a fill
        that started then). strict is the same read's date as a fresh check
        judges it. A None ev still dates the record."""
        with self._lock:
            if read_at is not None and (self._oldest is None or read_at < self._oldest):
                self._oldest = read_at
            if strict is not None and (self._strict is None or strict < self._strict):
                self._strict = strict
            if ev is None:
                return
            if len(self._items) - self._own >= MAX_CALLS:
                self._truncated = True
                return
            self._items.append(_Item(src=ev, by=by))

    def oldest(self) -> float | None:
        """When the oldest read behind the record began; None when every call
        was the check's own."""
        with self._lock:
            return self._oldest

    def oldest_strict(self) -> float | None:
        with self._lock:
            return self._strict

    def evidence(self) -> Evidence | None:
        """A snapshot of what was recorded, or None when nothing was."""
        with self._lock:
            if not self._items:
                return None
            return Evidence(list(self._items), self._truncated)


class _Key:
    __slots__ = ("name",)

    def __init__(self, name: str) -> None:
        self.name = name


_RECORDER = _Key("recorder")
_NO_RECORDER = object()
_FRESH = _Key("fresh")


def with_recorder(ctx: Context) -> tuple[Context, Recorder]:
    """A context carrying a new Recorder. The engine sets one up per check."""
    r = Recorder()
    return with_value(ctx, _RECORDER, r), r


def without_recorder(ctx: Context) -> Context:
    """A context whose recorder_from is None: auth runs on it and token
    fetches happen on it, since a token exchange is not what a decision was
    based on and its response carries the credential."""
    if ctx.value(_RECORDER) is None:
        return ctx
    return with_value(ctx, _RECORDER, _NO_RECORDER)


def recorder_from(ctx: Context | None) -> Recorder | None:
    if ctx is None:
        return None
    r = ctx.value(_RECORDER)
    return r if isinstance(r, Recorder) else None


def with_fresh(ctx: Context) -> Context:
    """Mark a context as belonging to a fresh check: every cache consulted
    under it looks the value up again and replaces its entry."""
    return with_value(ctx, _FRESH, True)


def fresh(ctx: Context | None) -> bool:
    return bool(ctx is not None and ctx.value(_FRESH))
